Added Viking Graben line12 example

kerim371 opened the pull request:
Viking Grabel line 12 is 2D seismic marine field data.

The added example contains preprocessing in Madagascar and FWI/RTM. LSRTM is in the process

mloubout commented:
Thanks a lot! Will look at it when back from holiday

rafaelorozco commented:
IM a student at SLIM lab. this PR and the PDF are so cool and informative. Thank you for sharing!

mloubout commented on examples/field_examples/viking_graben_line12/download_data.ipynb:
```diff
@@ -0,0 +1,146 @@
+{
```
Would make just a bash script for something like that

kerim371 replied:
ok

mloubout commented on examples/field_examples/viking_graben_line12/fwi/README.md:
```diff
@@ -0,0 +1,17 @@
+
+The idea is the following:
+
+1. prepare initial model
```
Start with line with capital letter. 

For 2 and 3, in practice that would not scale as you would have a massive data copy so you would probably just do "on-the-fly" filtering of a given shot record after reading it. Would be done using `judiFilter` in JUDI. How much work would it be to do that instead?

I also see that you use a slightly different filter than the one I have setup but could still wrap it in a preconditonner fairly easily

kerim371 replied:
Didn't know that. If `judiFilter` allows to apply the same filter to the source signal and shot records then it should work. I will try that.
One note is when we are trying to find minimal possible frequency we must be confident that the data (reflections/turnings waves) are still present after filtration. If we set low pass filter at 3 Hz for example then we may not see useful waves but only noise instead. Probably this can be added to example as a comment.

mloubout commented on examples/field_examples/viking_graben_line12/fwi/fwi.jl:
```diff
@@ -0,0 +1,278 @@
+    limit_m = true,
+    buffer_size = buffer_size,
+    optimal_checkpointing=false,
+    # subsampling_factor=t_sub,     # subsampling_factor with space_order set leads to exception (probably bug)
```
remove comments

kerim371 replied:
ok

mloubout commented on examples/field_examples/viking_graben_line12/rtm/rtm.jl:
```diff
@@ -0,0 +1,188 @@
+max_x = maximum([max_src_x, max_grp_x])
+
+# helpful functions
+rho_from_slowness(m) = 0.23.*(sqrt.(1f0 ./ m).*1000f0).^0.25
```
That's Gardner right? Maybe name it so

mloubout replied:
You can also put utilities that are used in multiple places (like fwi/rtm) in an `utils.jl`. DrWatson.jl can helo with that too.

kerim371 replied:
Right

mloubout commented on examples/field_examples/viking_graben_line12/fwi/fwi.jl:
```diff
@@ -0,0 +1,278 @@
+o = Tuple(Float32(i) for i in o)
+if modeling_type == "scalar"
+    model0 = Model(n, d, o, m0)
+elseif modeling_type == "acoustic"
```
Not sure about the `scalar/acoutic` name I would probably call it something like `slowness/bulk` maybe

mloubout commented on examples/field_examples/viking_graben_line12/fwi/fwi.jl:
```diff
@@ -0,0 +1,278 @@
+    # Subsampling the number of sources should in practice never be used for second order methods such as L-BFGS.
+    indsrc = shot_from:shot_step:shot_to
+    if mute_reflections
+        fval, gradient = fwi_objective(model0, q[indsrc], Ml_tur[indsrc]*d_obs[indsrc], options=jopt)
```
for filtering you would just have an extra `Ml_freq[indsrc]` in front of `q` and `d_obs`

mloubout commented on examples/field_examples/viking_graben_line12/fwi/fwi.jl:
```diff
@@ -0,0 +1,278 @@
+
+println("No.  ", "fval         ", "norm(gradient)")
+
+opt = Opt(:LD_LBFGS, prod(model0.n))
```
Any reason you chose this one over something like `pqn` that (in theory) handle constraints like bounds a bit better.

Just a question no need to change it

kerim371 replied:
Please share a link to `pqn`. Don't know about that.

mloubout commented on examples/field_examples/viking_graben_line12/fwi/README.md:
```diff
@@ -0,0 +1,17 @@
+# Full Waveform Inversion
```
nbviewer looks fine must bit github missbehaving

kerim371 replied:
I reran it.. Hope that helped

mloubout commented on examples/field_examples/viking_graben_line12/proc/README.md:
```diff
@@ -0,0 +1,18 @@
+# Viking Graben Line 12: processing
+
+The processing is done using [Madagascar](https://www.reproducibility.org/wiki/Main_Page) open source software. Thus we assume that Madagascar is present on the system.
```
This is quite nice. Too bas it can't be done in julia would have been a really nice package. So most of this is for RTM  and/or not used right?

mloubout replied:
Hum so there is a julia API

https://github.com/ahay/src/tree/master/api/julia

not sure if it actually works though

kerim371 replied:
Not sure if this is actually needed to rewrite it using Julia API. This will require some time to rework and Madagascar is very common to be used with Python. I understand that JUDI mostly relies on Julia but Python is required as well.

mloubout commented on examples/field_examples/viking_graben_line12/rtm/rtm.jl:
```diff
@@ -0,0 +1,188 @@
+# Right-hand preconditioners (model topmute)
+idx_wb = find_water_bottom(reshape(model0.m, model0.n))
+Tm = judiTopmute(model0.n, idx_wb, 10)  # Mute water column
+S = judiDepthScaling(model0)
```
There is a `judiIllumination` now in JUDI that might have been slightly better

kerim371 replied:
Don't how to use it yet. But interested in it.

mloubout commented on examples/field_examples/viking_graben_line12/rtm/rtm.jl:
```diff
@@ -0,0 +1,188 @@
+    dpi=600)
+Plots.savefig(rtm_plt, dir_out * "RTM_$(modeling_type).png")
+
+h5open(dir_out * "rtm_$(modeling_type).h5", "w") do file
```
could wrap this into a `write_model(file, name, model)` to be reused everywhere since the rest is mostly properties from `model`

kerim371 replied:
done

mloubout reviewed (commented):
Hi

Thanks so much for this contribution this is a great example and really happy to have a field data example. I think we should somehow add a page as well to the JUDI doc from this by maybe making a little markdown summary that shows the results and some steps. I.e I would remove the PDF and make instead a little markdown with those images, I will probably add the resulting RTM to the main page README.

I left some comments to try to make it more maintainable and reproducible, lemme know if you wanting to go that way and I'll try to help with it.

kerim371 commented on examples/field_examples/viking_graben_line12/README.md:
```diff
@@ -0,0 +1,19 @@
+
+It contains segy data, source signature, well logs and the description.
+
+Basically this example is dived in two parts: the processing (using Madagascar open source software) and inversion/migration (FWI/RTM/LSRTM) using `JUDI` interface.
```
You mean link to JUDI github repo right?

mloubout replied:
JUDI and Madagascar so people can find where/how to install it

kerim371 commented on examples/field_examples/viking_graben_line12/proc/README.md:
```diff
@@ -0,0 +1,18 @@
+
+The processing is done using [Madagascar](https://www.reproducibility.org/wiki/Main_Page) open source software. Thus we assume that Madagascar is present on the system.
+
+The processing is partly based on [Rodrigo Morelatto work](https://github.com/rmorel/Viking).
```
Deghosting and geometry are from mentioned repo. 
The original repo doesn' contain some important procedures like Radon filtration to supress multiples. It would be problematic to use original repo as is. Modifying original repo is also pretty big task at least because of name convention of processed files that I use.
I used it as a starting point.

kerim371 commented:
@mloubout generally I understood what and how  to do. In the nearest days I will try to fix that.

kerim371 commented:
@mloubout hi,

I'm trying to use `judiFilter` in that way:
```julia
Ml_freq = judiFilter(d_obs.geometry, 0.01, frq*1000f0)
Mr_freq = judiFilter(q.geometry, 0.01, frq*1000f0)

fval, gradient = fwi_objective(model0, Mr_freq[indsrc]*q[indsrc], Ml_freq[indsrc]*d_obs[indsrc], options=jopt)
```
but I get exception:
```
ERROR: MethodError: no method matching copy(::SeisCon)
Closest candidates are:
  copy(::Union{MathOptInterface.DualPowerCone, MathOptInterface.EqualTo, MathOptInterface.GreaterThan, MathOptInterface.Interval, MathOptInterface.LessThan, MathOptInterface.Parameter, MathOptInterface.PowerCone, MathOptInterface.Semicontinuous, MathOptInterface.Semiinteger}) at /home/kerim/Documents/Colada/r/julia-1.6/.julia/packages/MathOptInterface/4g9vU/src/sets.jl:2584
  copy(::Union{TransparentColor{C, T, N} where N, C} where {T, C<:Union{AbstractRGB{T}, AbstractGray{T}}}) at /home/kerim/Documents/Colada/r/julia-1.6/.julia/packages/ColorVectorSpace/QI5vM/src/ColorVectorSpace.jl:250
  copy(::DataStructures.Accumulator) at /home/kerim/Documents/Colada/r/julia-1.6/.julia/packages/DataStructures/59MD0/src/accumulator.jl:41
  ...
Stacktrace:
  [1] *(n::Float32, s::SeisCon)
    @ JUDI ~/Documents/Colada/r/julia-1.6/.julia/dev/JUDI/src/TimeModeling/Types/judiVector.jl:156
  [2] _broadcast_getindex_evalf
    @ ./broadcast.jl:648 [inlined]
  [3] _broadcast_getindex
    @ ./broadcast.jl:621 [inlined]
  [4] getindex
    @ ./broadcast.jl:575 [inlined]
  [5] copy
    @ ./broadcast.jl:922 [inlined]
  [6] materialize
    @ ./broadcast.jl:883 [inlined]
  [7] zero(::Type{Float32}, v::judiVector{Float32, SeisCon}; nsrc::Int64)
    @ JUDI ~/Documents/Colada/r/julia-1.6/.julia/dev/JUDI/src/TimeModeling/Types/judiVector.jl:137
  [8] zero(::Type{Float32}, v::judiVector{Float32, SeisCon})
    @ JUDI ~/Documents/Colada/r/julia-1.6/.julia/dev/JUDI/src/TimeModeling/Types/judiVector.jl:137
  [9] copy
    @ ~/Documents/Colada/r/julia-1.6/.julia/dev/JUDI/src/TimeModeling/Types/abstract.jl:36 [inlined]
 [10] deepcopy
    @ ~/Documents/Colada/r/julia-1.6/.julia/dev/JUDI/src/TimeModeling/Types/abstract.jl:37 [inlined]
 [11] matvec(#unused#::FrequencyFilter{Float32, 0.01f0, 5.0f0}, Din::judiVector{Float32, SeisCon})
    @ JUDI ~/Documents/Colada/r/julia-1.6/.julia/dev/JUDI/src/TimeModeling/Preconditioners/DataPreconditioners.jl:158
 [12] *(J::FrequencyFilter{Float32, 0.01f0, 5.0f0}, ms::judiVector{Float32, SeisCon})
    @ JUDI ~/Documents/Colada/r/julia-1.6/.julia/dev/JUDI/src/TimeModeling/Preconditioners/base.jl:28
 [13] top-level scope
    @ REPL[2]:1
```

It seems `SeisCon` is not supported. 
Probably `SeisBlock` is supported but I prefer not to load all the SEGY to RAM with `SeisBlock`.
May we fix that?

kerim371 commented:
@mloubout hi,

Should I wait for `judiFilter` exception to be fixed so I could continue working on the PR?

mloubout commented:
>Should I wait for judiFilter exception to be fixed so I could continue working on the PR?

You can do `F*get_data(dobs)` to convert it to in-core for the time being.
For the frequencies, just put something very low for fmin (like fmin=.001) to do low-pass

kerim371 commented:
Tested with FWI and RTM should work too.

mloubout commented:
Ok I think it's mostly there. Can you
- Rerun the notebooks, it looks like some of the preprocessing figures are broken (the degosting for example)
- Remove all `Pkg.add` `pip install` those should be in a setup script (or idealy a Project/Manifest.toml)
- Remove all commented lines of code (i.e `# Pkg.add ...`)
- I still think should only keep the Madagascar notebook used for this example, and you can add a links to that other repo for more information.


The final step would be to add some of the notebooks/md files to the documentation. But I can do that myself once this is merged

kerim371 commented:
@mloubout hi,

It seems I cleaned it all up.

It seems github is able visualize `matplotlib` images without problems. But `plotly` wasn't working that is why in `deghosting` notebook we could not see some pictures.
I modified `proc` notebooks so they work with `matplotlib` only.

Most likely the same problem we encounter when looking at `initial_model` notebook which uses Julia's `Plots` library for plotting. 
I guess if I use PyCall's `matplotlib` it will work fine. How do you think should I try?

> * I still think should only keep the Madagascar notebook used for this example, and you can add a links to that other repo for more information.

If you don't mind I suggest to ask opinion/permission the author of that repo @rmorel.

Basically two steps from preprocessing (geometry check and deghosting) of Viking Graben line 12 are based on @rmorel [work](https://github.com/rmorel/Viking).

If I simply use this package then it will be unsufficient.
If we recommend the user to preprocess the data using original repo and then come back to this processing flow then it will be confusing. Moreover the original repo is a little old and may not work properly with newer version of pandas. 

I think the processing should not be divided in two packages. So I suggest either leave it as is or improve the original repo probably with my additions.

rmorel commented:
@kerim371, all the code and work in my [repository](https://github.com/rmorel/Viking) is open source (GPL). Feel free to reuse or modify the code in any way you want.

I'm happy to hear that it's useful =).

mloubout commented:
>I think the processing should not be divided in two packages.

Sorry wasn't what I mean. I only meant to remove the notebooks that do processing steps not sued for  the FWI/RTM so that this example is self-contained.

Thanks for the plots. Sorry I'm a bit busy these days but not forgetting about this great addition.


Would you mind If I pushed some changes to this PR myself. I won't change any of the actual content but I want to set it up with `DrWatson.jl` so that it is safely fully reproducible at least for the julia side of it.

kerim371 commented:
> Sorry wasn't what I mean. I only meant to remove the notebooks that do processing steps not sued for the FWI/RTM so that this example is self-contained.

If I understood you then processing notebooks are required for RTM/LSRTM.
FWI doesn't require these notebooks.
So you mean completely remove these processing notebooks from PR (for example create another repository on github with these notebooks) or restructure the folders so the user could understand that FWI doesn't require processing?



> Would you mind If I pushed some changes to this PR myself. I won't change any of the actual content but I want to set it up with `DrWatson.jl` so that it is safely fully reproducible at least for the julia side of it.

Yes of course.

Also there are few things that I'm working on: 
1. try to replace julia's Plots by PyPlot in `initial_model` notebook so that the images were visible on github
2. use `ImageGathers` for `QC` and further try to perform AVO
3. add LSRTM script

I don't know when I finish these so you can merge it whenever you want. 
If I succeeded then I will reopen/crete new PR with these changes.

kerim371 commented:
Hi @mloubout,

I started to explore `pqn` optimization in hope to achieve the better results.
For now it seems that it can work better but I don't know what parameters should I try to play.
It would be good if you give me advice what should I try change.

The result I get for 3.5 Hz:
![image](https://github.com/slimgroup/JUDI.jl/assets/43808863/dda66ab2-bc1c-4ec7-8fb9-33862dd32925)
![image](https://github.com/slimgroup/JUDI.jl/assets/43808863/04b7a0aa-03db-408d-acfb-f9c51f0063ad)
![image](https://github.com/slimgroup/JUDI.jl/assets/43808863/6e40b8e7-8d3d-47c9-ac2a-1bd0199c2443)

mloubout commented:
@kerim371 would you mind checking if this still runs properly with the new changes. I set up this project with DrWatson.jl so that all packages used are fixed at a given version making it fully reproducible.

kerim371 commented:
@mloubout I made some small fixes. 

I haven't worked with `DrWatson` until now and this is how I tried to use this field example:
1. `] dev JUDI`
2. follow instructions in `README.md` like: preinstall Julia deps (along with `DrWatson.jl`)
3. run script `fw_lbfgs.jl`

If that is the way how users are going to use it now then it should work fine.
